File: leetcode/spiral-matrix.java
//https://leetcode.com/problems/spiral-matrix/

import java.util.ArrayList;
import java.util.List;

class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        ArrayList<Integer> spiral = new ArrayList<Integer>();
        int rowsVisited = 0;
        int colsVisited = 0;
                
        while(rowsVisited < matrix.length && colsVisited < matrix[0].length){               
            //right
            for(int i = colsVisited; i<matrix[0].length-colsVisited; i++){
                spiral.add(matrix[rowsVisited][i]);
            }
            rowsVisited++;
            if(matrix.length*matrix[0].length == spiral.size()) break;
            //down
            for(int i = rowsVisited; i<matrix.length-(rowsVisited-1); i++){
                spiral.add(matrix[i][matrix[0].length - 1 - colsVisited]);
            }
            colsVisited++;
            if(matrix.length*matrix[0].length == spiral.size()) break;
            //left
            for(int i = matrix[0].length-colsVisited-1; i>=colsVisited-1; i--){
                spiral.add(matrix[matrix.length-rowsVisited][i]);
            }
            if(matrix.length*matrix[0].length == spiral.size()) break;
            //up
            for(int i = matrix.length-rowsVisited-1; i>=rowsVisited; i--){
                spiral.add(matrix[i][colsVisited-1]);
            }
            if(matrix.length*matrix[0].length == spiral.size()) break;
        }
        
        return spiral;
    }
}

File: leetcode/spiral_matrix.py
# https://leetcode.com/problems/spiral-matrix/description/


class Solution:
    def spiralOrder(self, matrix: list[list[int]]) -> list[int]:
        row: int = 0
        col: int = 0
        order: list[int] = []

        while len(order) < len(matrix) * len(matrix[0]):
            # go right
            for c in range(col, len(matrix[0]) - col):
                if len(order) < len(matrix) * len(matrix[0]):
                    order.append(matrix[row][c])

            # go down
            for r in range(row + 1, len(matrix) - row):
                if len(order) < len(matrix) * len(matrix[0]):
                    order.append(matrix[r][len(matrix[0]) - 1 - col])

            # go left
            for c in range(len(matrix[0]) - col - 2, col, -1):
                if len(order) < len(matrix) * len(matrix[0]):
                    order.append(matrix[len(matrix) - 1 - row][c])

            # go up
            for r in range(len(matrix) - row - 1, row, -1):
                if len(order) < len(matrix) * len(matrix[0]):
                    order.append(matrix[r][col])

            col += 1
            row += 1

        return order
